graph/src/schema/input_schema.rs

```diff
@@ -395,6 +435,332 @@ impl EnumMap {
+        let source = self.aggregation(schema).source;
+        EntityType::new(schema.cheap_clone(), source)
+    }
+
+    pub fn aggregation<'a>(&self, schema: &'a InputSchema) -> &'a Aggregation {
+        schema.inner.type_infos[self.aggregation]
+            .aggregation()
+            .expect("the aggregation source is an object types")
+    }
+
+    pub fn agg_type(&self, schema: &InputSchema) -> EntityType {
+        let agg_type = self.aggregation(schema).obj_types[self.agg_type].name;
+        EntityType::new(schema.cheap_clone(), agg_type)
+    }
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct Arg {
+    pub name: Word,
+    pub value_type: ValueType,
+}
+
+impl Arg {
+    fn new(name: Word, src_type: &s::ObjectType) -> Self {
+        let value_type = src_type
+            .field(&name)
+            .unwrap()
+            .field_type
+            .value_type()
+            .unwrap();
+        Self { name, value_type }
+    }
+}
+#[derive(PartialEq, Debug)]
+pub struct Aggregate {
+    pub name: Word,
+    pub func: AggregateFn,
+    pub arg: Option<Arg>,
+    pub field_type: s::Type,
+    pub value_type: ValueType,
+}
+
+impl Aggregate {
+    fn new(
+        _schema: &Schema,
+        src_type: &s::ObjectType,
+        name: &str,
+        field_type: &s::Type,
+        dir: &s::Directive,
+    ) -> Self {
+        let func = dir
+            .argument("fn")
+            .unwrap()
+            .as_str()
+            .unwrap()
+            .parse()
+            .unwrap();
+        let arg = dir
+            .argument("arg")
+            .map(|arg| Word::from(arg.as_str().unwrap()))
+            .map(|arg| Arg::new(arg, src_type));
+        Aggregate {
+            name: Word::from(name),
+            func,
+            arg,
+            field_type: field_type.clone(),
+            value_type: field_type.get_base_type().parse().unwrap(),
+        }
+    }
+
+    /// The field needed for the finalised aggregation for hourly/daily
+    /// values
+    fn as_agg_field(&self) -> Field {
+        Field {
+            name: self.name.clone(),
+            field_type: self.field_type.clone(),
+            value_type: self.value_type,
+            is_derived: false,
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Aggregation {
+    pub name: Atom,
+    pub id_type: IdType,
+    pub intervals: Box<[AggregationInterval]>,
+    pub source: Atom,
+    /// The non-aggregation fields of the time series
+    pub fields: Box<[Field]>,
+    pub aggregates: Box<[Aggregate]>,
+    /// The object types for the aggregated data, one for each interval, in
+    /// the same order as `intervals`
+    obj_types: Box<[ObjectType]>,
+}
+
+impl Aggregation {
+    pub fn new(schema: &Schema, pool: &AtomPool, agg_type: &s::ObjectType) -> Self {
+        let name = pool.lookup(&agg_type.name).unwrap();
+        let id_type = IdType::try_from(agg_type).expect("validation caught any issues here");
```

Comment:
Should this return an error instead?

Reply:
No, we should never get here since validations reject a subgraph where the type of `id` would fail this conversion. The problem is that we first do validations, and then extract things from the validated AST. It would be much better to do the extraction at the same time as validation, but that's a pretty big change that I didn't want to also shoehorn into this PR.

A big part of this and a couple previous PR's is to isolate these kinds of conversions by only doing them while we construct the `InputSchema`; previously, a lot of this happened everywhere where the AST was used (since there were no internal data structures to represent our idea of schema information)